src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs

```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+                        if (!visitor(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Right.Kind);
+                }
+            }
+
+            Debug.Assert(stack.Count == 0);
+            stack.Free();
+            return true;
+
+            static void pushLeftNodes(BoundBinaryOperator binary, ArrayBuilder<BoundBinaryOperator> stack)
+            {
+                Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString) || binary.IsUnconvertedInterpolatedStringAddition);
+                BoundBinaryOperator? current = binary;
+                while (current != null)
+                {
+                    stack.Push(current);
+                    current = current.Left as BoundBinaryOperator;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Rewrites a BoundBinaryOperator composed of interpolated strings (either converted or unconverted) iteratively, without
+        /// recursion.
+        /// </summary>
+        /// <param name="binary">The original top of the binary operations.</param>
+        /// <param name="arg">The callback args.</param>
+        /// <param name="interpolatedStringFactory">
+        /// Rewriter for the BoundInterpolatedString or BoundUnconvertedInterpolatedString parts of the binary operator. Passed the callback
+        /// parameter, the original interpolated string, and the index of the interpolated string in the tree.
+        /// </param>
+        /// <param name="binaryOperatorFactory">
+        /// Rewriter for the BoundBinaryOperator parts fo the binary operator. Passed the callback parameter, the original binary operator, and
+        /// the rewritten left and right components.
+        /// </param>
+        public static TResult RewriteInterpolatedStringAddition<TArg, TInterpolatedStringType, TResult>(
+            this BoundBinaryOperator binary,
+            TArg arg,
+            Func<TArg, TInterpolatedStringType, int, TResult> interpolatedStringFactory,
+            Func<TArg, BoundBinaryOperator, TResult, TResult, TResult> binaryOperatorFactory)
+            where TInterpolatedStringType : BoundExpression
+        {
+            Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundUnconvertedInterpolatedString) || typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString));
+            var originalStack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+            var rewrittenLefts = ArrayBuilder<(BoundExpression original, TResult rewritten)>.GetInstance();
+            (BoundBinaryOperator? original, TResult? rewritten) result = default;
+
+            pushLeftNodes(binary, originalStack);
+
+            int i = 0;
+            while (originalStack.TryPeek(out var currentBinary))
```

Comment:
>while (originalStack.TryPeek(out var currentBinary)) [](http://example.com/codeflow?start=12&length=52)

It feels like the implementation (if correct) is too complicated and therefore the correctness is hard to prove. Let's discuss offline ways to simplify the implementation.